File: sdk/clojure/src/rocksky/remote.clj
(ns rocksky.remote
  "Remote-control bindings: build a Rocksky-controllable player or a remote UI.

  Two roles over the remote-control WebSocket (remote-ws/PROTOCOL.md), both
  wrapping the shared Rust core's C ABI (crates/rocksky-uniffi) via JVM Panama
  FFM — the same opaque-handle style as `rocksky.core`'s agent:

    - a RemotePlayer advertises now-playing/status/queue and receives commands;
    - a RemoteController lists devices, picks the primary, and sends commands.

  Both use a poll model: a blocking `next-command`/`next-event` you loop on a
  background thread. `listen` does that for you — pass a handler map and it spawns
  a thread that dispatches parsed commands/events to your fns.

    (def player (remote/connect-player token \"My Player\"))
    (remote/listen player {:play  #(engine-play)
                           :pause #(engine-pause)
                           :seek  (fn [ms] (engine-seek ms))})
    (remote/set-now-playing player {:title \"…\" :artist \"…\" :durationMs 200000
                                    :elapsedMs 0 :isPlaying true})
    (remote/set-status player :playing)

    (def ctl (remote/connect-controller token \"My Controller\"))
    (remote/listen ctl {:devices     (fn [e] (render (get e \"devices\")))
                        :now-playing (fn [e] (show (get e \"track\")))})
    (remote/set-primary ctl device-id)
    (remote/pause ctl device-id)       ; nil/omitted target ⇒ broadcast

  Records/commands/events are Clojure maps with camelCase string keys — the same
  convention as `rocksky.core`. Requires JDK 22+ (run with
  --enable-native-access=ALL-UNNAMED) and the native lib built with the
  `remote-player` feature (the crate default)."
  (:refer-clojure :exclude [next])
  (:require [cheshire.core :as json]
            [clojure.string :as str]
            [rocksky.native :as native])
  (:import [java.lang.foreign Arena Linker FunctionDescriptor SymbolLookup
            ValueLayout MemoryLayout MemorySegment Linker$Option]
           [java.lang.invoke MethodHandle]))

;; Native resolution is deferred behind delays so merely requiring this namespace
;; touches no native code; the library is resolved on the first actual call.
(def ^:private lib-path (delay (native/resolve-lib)))
(def ^:private arena  (delay (Arena/ofShared)))
(def ^:private linker (delay (Linker/nativeLinker)))
(def ^:private lookup (delay (SymbolLookup/libraryLookup @lib-path ^Arena @arena)))

(def ^:private ADDR ValueLayout/ADDRESS)
(def ^:private I32  ValueLayout/JAVA_INT)     ; uint32_t
(def ^:private I64  ValueLayout/JAVA_LONG)    ; uint64_t
(def ^:private BOOL ValueLayout/JAVA_BOOLEAN) ; C _Bool (1 byte)

(defn- ^MethodHandle downcall
  "Bind a C function `name` to a MethodHandle. `ret` is nil for void."
  [name ret arg-layouts]
  (let [seg (.orElseThrow (.find ^SymbolLookup @lookup name))
        arr (into-array MemoryLayout arg-layouts)
        fd  (if ret
              (FunctionDescriptor/of ret arr)
              (FunctionDescriptor/ofVoid arr))]
    (.downcallHandle ^Linker @linker seg fd (make-array Linker$Option 0))))

(def ^:private h-free (delay (downcall "rocksky_string_free" nil [ADDR])))

;; ---- player handles -----------------------------------------------------
(def ^:private h-p-connect     (delay (downcall "rocksky_remote_player_connect" ADDR [ADDR ADDR ADDR])))
(def ^:private h-p-next        (delay (downcall "rocksky_remote_player_next_command" ADDR [ADDR])))
(def ^:private h-p-nowplaying  (delay (downcall "rocksky_remote_player_set_now_playing" ADDR [ADDR ADDR])))
(def ^:private h-p-status      (delay (downcall "rocksky_remote_player_set_status" ADDR [ADDR ADDR])))
(def ^:private h-p-queue       (delay (downcall "rocksky_remote_player_set_queue" ADDR [ADDR ADDR I32])))
(def ^:private h-p-disconnect  (delay (downcall "rocksky_remote_player_disconnect" ADDR [ADDR])))
(def ^:private h-p-free        (delay (downcall "rocksky_remote_player_free" nil [ADDR])))

;; ---- controller handles -------------------------------------------------
(def ^:private h-c-connect     (delay (downcall "rocksky_remote_controller_connect" ADDR [ADDR ADDR ADDR])))
(def ^:private h-c-next        (delay (downcall "rocksky_remote_controller_next_event" ADDR [ADDR])))
(def ^:private h-c-primary     (delay (downcall "rocksky_remote_controller_set_primary" ADDR [ADDR ADDR])))
(def ^:private h-c-command     (delay (downcall "rocksky_remote_controller_command" ADDR [ADDR ADDR ADDR])))
(def ^:private h-c-seek        (delay (downcall "rocksky_remote_controller_seek" ADDR [ADDR ADDR I64])))
(def ^:private h-c-qjump       (delay (downcall "rocksky_remote_controller_queue_jump" ADDR [ADDR ADDR I32])))
(def ^:private h-c-qremove     (delay (downcall "rocksky_remote_controller_queue_remove" ADDR [ADDR ADDR I32])))
(def ^:private h-c-enqueue     (delay (downcall "rocksky_remote_controller_enqueue" ADDR [ADDR ADDR ADDR ADDR BOOL I32])))
(def ^:private h-c-disconnect  (delay (downcall "rocksky_remote_controller_disconnect" ADDR [ADDR])))
(def ^:private h-c-free        (delay (downcall "rocksky_remote_controller_free" nil [ADDR])))

(defn- read-free
  "Read an owned C string returned by the core, then free it. nil on NULL."
  [^MemorySegment seg]
  (when (and seg (not (zero? (.address seg))))
    (let [s (.getString (.reinterpret seg Long/MAX_VALUE) 0)]
      (.invokeWithArguments ^MethodHandle @h-free (object-array [seg]))
      s)))

(defn- unwrap
  "Parse a {\"ok\"|\"error\"} envelope, throwing ex-info on error."
  [seg]
  (let [m (json/parse-string (read-free seg))]
    (if (contains? m "error")
      (throw (ex-info (str "rocksky: " (get m "error")) {:error (get m "error")}))
      (get m "ok"))))

(defn- cstr
  "Allocate a NUL-terminated UTF-8 C string in `a` (nil ⇒ empty string)."
  [^Arena a s]
  (.allocateFrom a (str (or s ""))))

;; ---- lifecycle protocol -------------------------------------------------
;;
;; A player and a controller are opaque native handles wrapped in a record.
;; They share the same lifecycle verbs, so those live on a protocol; the
;; role-specific verbs are plain functions below.

(defprotocol Remote
  (disconnect [this]
    "Disconnect and stop the background task. Unblocks the `listen` loop (its
    next poll returns nil, ending the thread). The handle stays valid until
    `close`.")
  (close [this]
    "Release the native handle. Call once, after `disconnect`.")
  (listen [this handlers]
    "Spawn a background thread that loops the blocking poll and dispatches to
    `handlers` (a map of keyword ⇒ fn). Returns the `future`; stop it with
    `disconnect`. An optional `:error` handler receives any Throwable."))

(defrecord RemotePlayer [^MemorySegment handle])
(defrecord RemoteController [^MemorySegment handle])

;; ---- player -------------------------------------------------------------

(defn connect-player
  "Connect and register a controllable player. `token` is a Rocksky access token
  (JWT); `name` is the device-picker label; `url` may be nil for the public
  endpoint. Returns a `RemotePlayer`."
  ([token name] (connect-player token name nil))
  ([token name url]
   (with-open [^Arena a (Arena/ofConfined)]
     (let [^MemorySegment seg
           (.invokeWithArguments ^MethodHandle @h-p-connect
                                 (object-array [(cstr a token) (cstr a name) (cstr a url)]))]
       (->RemotePlayer seg)))))

(defn set-now-playing
  "Advertise the current track. `track` is a map with camelCase keys: :title,
  :artist, :album, :albumArtist, :albumArt, :durationMs, :elapsedMs, :isPlaying,
  plus optional audio info :codec (audio codec/container, e.g. \"mp3\", \"flac\")
  and :sampleRate (sample rate in Hz, e.g. 44100), included in the track payload
  when set and omitted otherwise."
  [^RemotePlayer player track]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-p-nowplaying
                                  (object-array [(:handle player)
                                                 (cstr a (json/generate-string track))])))))

(defn set-status
  "Advertise transport state — :playing, :paused, or :stopped (keyword or string)."
  [^RemotePlayer player status]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-p-status
                                  (object-array [(:handle player) (cstr a (name status))])))))

(defn set-queue
  "Advertise the playback queue + active index. `items` is a vector of queue-item
  maps (camelCase keys: :uploadId/:trackId, :title, :artist, :album,
  :albumArtist, :albumArt, :durationMs, :songUri, :albumUri, :trackNumber)."
  [^RemotePlayer player items index]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-p-queue
                                  (object-array [(:handle player)
                                                 (cstr a (json/generate-string (or items [])))
                                                 (int index)])))))

(defn next-command
  "Block until the next controller command; returns a parsed command map (string
  keys) or nil once disconnected. Prefer `listen` — this is the raw poll."
  [^RemotePlayer player]
  (unwrap (.invokeWithArguments ^MethodHandle @h-p-next (object-array [(:handle player)]))))

(defn- dispatch-command
  "Route one parsed command to the matching handler in `handlers`."
  [handlers cmd]
  (let [action (get cmd "action")
        k (keyword (str/replace (str action) "_" "-"))]
    (when-let [h (get handlers k)]
      (case (str action)
        ("play" "pause" "next" "previous") (h)
        "seek"          (h (get cmd "position"))
        "queue_jump"    (h (get cmd "index"))
        "queue_remove"  (h (get cmd "index"))
        "enqueue"       (h {:tracks     (get cmd "tracks")
                            :mode       (get cmd "mode")
                            :shuffle    (get cmd "shuffle")
                            :startIndex (get cmd "startIndex")})
        (h cmd)))))

(defn- player-loop [player handlers]
  (loop []
    (when-let [cmd (try (next-command player)
                        (catch Throwable t
                          (when-let [e (:error handlers)] (e t)) nil))]
      (try (dispatch-command handlers cmd)
           (catch Throwable t (when-let [e (:error handlers)] (e t))))
      (recur))))

;; ---- controller ---------------------------------------------------------

(defn connect-controller
  "Connect and register a controller (remote UI). Args as `connect-player`.
  Returns a `RemoteController`."
  ([token name] (connect-controller token name nil))
  ([token name url]
   (with-open [^Arena a (Arena/ofConfined)]
     (let [^MemorySegment seg
           (.invokeWithArguments ^MethodHandle @h-c-connect
                                 (object-array [(cstr a token) (cstr a name) (cstr a url)]))]
       (->RemoteController seg)))))

(defn set-primary
  "Choose the primary (scrobble/profile) device by id."
  [^RemoteController controller device-id]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-c-primary
                                  (object-array [(:handle controller) (cstr a device-id)])))))

(defn- command
  "Send a simple command (play|pause|next|previous). nil/empty `target` ⇒ broadcast."
  [^RemoteController controller action target]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-c-command
                                  (object-array [(:handle controller)
                                                 (cstr a (name action))
                                                 (cstr a target)])))))

(defn play
  "Play on `target` (a device id), or omit/nil to broadcast to all devices."
  ([controller] (play controller nil))
  ([controller target] (command controller "play" target)))

(defn pause
  ([controller] (pause controller nil))
  ([controller target] (command controller "pause" target)))

(defn next
  ([controller] (next controller nil))
  ([controller target] (command controller "next" target)))

(defn previous
  ([controller] (previous controller nil))
  ([controller target] (command controller "previous" target)))

(defn seek
  "Seek `target` to `position-ms` (nil target ⇒ broadcast)."
  [^RemoteController controller target position-ms]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-c-seek
                                  (object-array [(:handle controller)
                                                 (cstr a target)
                                                 (long position-ms)])))))

(defn queue-jump
  "Jump to `index` in `target`'s queue (nil target ⇒ broadcast)."
  [^RemoteController controller target index]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-c-qjump
                                  (object-array [(:handle controller)
                                                 (cstr a target)
                                                 (int index)])))))

(defn queue-remove
  "Remove queue item `index` on `target` (nil target ⇒ broadcast)."
  [^RemoteController controller target index]
  (with-open [^Arena a (Arena/ofConfined)]
    (unwrap (.invokeWithArguments ^MethodHandle @h-c-qremove
                                  (object-array [(:handle controller)
                                                 (cstr a target)
                                                 (int index)])))))

(defn enqueue
  "Enqueue `tracks` (a vector of queue-item maps) on `target` (nil ⇒ broadcast).
  `mode` is :now | :next | :last (default :now); `shuffle` a bool (default false);
  `start-index` an int (default 0)."
  ([controller target tracks] (enqueue controller target tracks :now false 0))
  ([^RemoteController controller target tracks mode shuffle start-index]
   (with-open [^Arena a (Arena/ofConfined)]
     (unwrap (.invokeWithArguments ^MethodHandle @h-c-enqueue
                                   (object-array [(:handle controller)
                                                  (cstr a target)
                                                  (cstr a (json/generate-string (or tracks [])))
                                                  (cstr a (name mode))
                                                  (boolean shuffle)
                                                  (int start-index)]))))))

(defn next-event
  "Block until the next update; returns a parsed event map (string keys) or nil
  once disconnected. Prefer `listen` — this is the raw poll."
  [^RemoteController controller]
  (unwrap (.invokeWithArguments ^MethodHandle @h-c-next (object-array [(:handle controller)]))))

(defn- dispatch-event
  "Route one parsed event to the handler keyed by its `type`
  (:devices :device-registered :device-unregistered :primary-changed
  :now-playing :status :queue). The handler receives the whole event map. A
  :now-playing event's \"track\" map includes optional \"codec\" / \"sampleRate\"
  audio info when the player advertised it."
  [handlers ev]
  (let [k (keyword (str/replace (str (get ev "type")) "_" "-"))]
    (when-let [h (get handlers k)]
      (h ev))))

(defn- controller-loop [controller handlers]
  (loop []
    (when-let [ev (try (next-event controller)
                       (catch Throwable t
                         (when-let [e (:error handlers)] (e t)) nil))]
      (try (dispatch-event handlers ev)
           (catch Throwable t (when-let [e (:error handlers)] (e t))))
      (recur))))

;; ---- lifecycle impls ----------------------------------------------------

(extend-protocol Remote
  RemotePlayer
  (disconnect [this]
    (unwrap (.invokeWithArguments ^MethodHandle @h-p-disconnect (object-array [(:handle this)]))))
  (close [this]
    (.invokeWithArguments ^MethodHandle @h-p-free (object-array [(:handle this)]))
    nil)
  (listen [this handlers]
    (future (player-loop this handlers)))

  RemoteController
  (disconnect [this]
    (unwrap (.invokeWithArguments ^MethodHandle @h-c-disconnect (object-array [(:handle this)]))))
  (close [this]
    (.invokeWithArguments ^MethodHandle @h-c-free (object-array [(:handle this)]))
    nil)
  (listen [this handlers]
    (future (controller-loop this handlers))))

File: sdk/ruby/lib/rocksky/remote.rb
# frozen_string_literal: true

module Rocksky
  # RemotePlayer — build a Rocksky-controllable player in a few lines.
  #
  # It speaks the remote-control WebSocket protocol (see remote-ws/PROTOCOL.md)
  # over the shared Rust core: it registers as a device, advertises what you're
  # playing (now-playing / status / queue), and invokes your handlers when a
  # miniplayer sends a command (play/pause/next/previous/seek/enqueue/queue
  # actions). Heartbeat, reconnect, and the device-id handshake are handled for
  # you inside the core.
  #
  # Commands arrive over a blocking poll (#next_command). Register handlers with
  # #on and call #listen to spawn a background thread that dispatches them, so you
  # never write the loop yourself.
  #
  #   player = Rocksky::RemotePlayer.connect(token, "My Player")
  #   player.on(:play)  { engine.play }
  #   player.on(:pause) { engine.pause }
  #   player.on(:seek)  { |cmd| engine.seek(cmd[:position]) }
  #   player.listen
  #   # …then, as your engine plays:
  #   player.set_now_playing(title: "Chaser", artist: "Calibro 35",
  #                          durationMs: 182_320, elapsedMs: 0, isPlaying: true)
  #   player.set_status("playing")
  #   player.set_queue(items, 0)
  #
  # Records (now-playing / queue items) are passed as Hashes with camelCase keys
  # (title, artist, album, albumArtist, albumArt, durationMs, elapsedMs,
  # isPlaying, codec, sampleRate / songUri, albumUri, trackNumber), matching the
  # wire record shape.
  class RemotePlayer
    # Connect and register a controllable player in the background. +token+ is a
    # Rocksky access token (JWT); +name+ is the device-picker label; +url+
    # defaults to the public endpoint when omitted.
    def self.connect(token, name, url: nil)
      ptr = C.rocksky_remote_player_connect(token.to_s, name.to_s, url.to_s)
      raise Error, (Rocksky.take_string(C.rocksky_last_error()) || "connect failed") if ptr.null?

      new(ptr)
    end

    def initialize(ptr)
      @ptr = ptr
      @handlers = {}
      @thread = nil
    end

    # Register a handler for a command action, one of +:play+, +:pause+, +:next+,
    # +:previous+, +:seek+, +:queue_jump+, +:queue_remove+, +:queue_move+,
    # +:enqueue+, +:shuffle+, +:repeat+, +:volume+, +:audio_settings+. The block
    # receives the full symbol-keyed command Hash (e.g. +{ action: "seek",
    # position: 42000 }+). Returns +self+ for chaining.
    def on(action, &block)
      @handlers[action.to_sym] = block
      self
    end

    # Block until the next controller command, returned as a symbol-keyed Hash
    # (e.g. +{ action: "play" }+), or +nil+ once disconnected.
    def next_command
      return nil if @ptr.null?

      cmd = Rocksky.unwrap(C.rocksky_remote_player_next_command(@ptr))
      cmd && Rocksky.symbolize(cmd)
    end

    # Spawn a background thread that polls #next_command and dispatches each
    # command to its registered handler until the player disconnects. Returns the
    # Thread.
    def listen
      @thread ||= Thread.new do
        loop do
          cmd = next_command
          break if cmd.nil?

          dispatch(cmd)
        end
      end
    end

    # Advertise the currently-playing track. Call whenever it changes, and
    # periodically (~every 1–4s) with a fresh +elapsedMs+ so controllers show
    # smooth progress. +track+ is a camelCase Hash. Optional audio info —
    # +codec+ (audio codec/container, e.g. "mp3", "flac") and +sampleRate+
    # (sample rate in Hz, e.g. 44100) — is included in the track payload when
    # set, and omitted otherwise.
    def set_now_playing(track)
      Rocksky.unwrap(C.rocksky_remote_player_set_now_playing(@ptr, JSON.generate(track)))
    end

    # Advertise transport state: "playing", "paused", or "stopped".
    def set_status(status)
      Rocksky.unwrap(C.rocksky_remote_player_set_status(@ptr, status.to_s))
    end

    # Advertise the playback queue (+items+ is an array of camelCase queue-item
    # Hashes) and the active +index+.
    def set_queue(items, index)
      Rocksky.unwrap(C.rocksky_remote_player_set_queue(@ptr, JSON.generate(items), index))
    end

    # Disconnect and stop the background task. #next_command then returns nil,
    # which ends the #listen thread. The handle stays valid until #close.
    def disconnect
      Rocksky.unwrap(C.rocksky_remote_player_disconnect(@ptr))
    end

    # Release the native handle. The player is unusable afterwards.
    def close
      return if @ptr.null?

      # Stop the background task so a blocked #next_command returns nil and the
      # #listen loop exits, then WAIT for that thread to leave the native call
      # before freeing — freeing the handle while a poll is in flight is a
      # use-after-free (segfault).
      C.rocksky_remote_player_disconnect(@ptr)
      @thread&.join unless @thread == Thread.current
      @thread = nil
      C.rocksky_remote_player_free(@ptr)
      @ptr = Fiddle::Pointer.new(0)
    end

    private

    def dispatch(cmd)
      handler = @handlers[cmd[:action]&.to_sym]
      handler&.call(cmd)
    end
  end

  # RemoteController — build a Rocksky remote UI in a few lines.
  #
  # The other half of the remote-control WebSocket protocol: where
  # {RemotePlayer} is a controllable player, a controller lists the user's
  # players, observes what each is playing (now-playing / status / queue), picks
  # the primary device, and sends commands. Heartbeat, reconnect, and the
  # register handshake are handled for you inside the core.
  #
  # Events arrive over a blocking poll (#next_event). Register handlers with #on
  # and call #listen to spawn a background dispatch thread.
  #
  #   controller = Rocksky::RemoteController.connect(token, "My Controller")
  #   controller.on(:devices)     { |e| render(e[:devices]) }
  #   controller.on(:now_playing) { |e| show_track(e[:deviceId], e[:track]) }
  #   controller.listen
  #   # …then drive a device:
  #   controller.set_primary(device_id)
  #   controller.pause(device_id)
  #   controller.seek(device_id, 42_000)
  #
  # A +nil+ / omitted +target+ broadcasts a command to all of the user's devices.
  class RemoteController
    # Connect and register a controller in the background. +url+ defaults to the
    # public endpoint when omitted.
    def self.connect(token, name, url: nil)
      ptr = C.rocksky_remote_controller_connect(token.to_s, name.to_s, url.to_s)
      raise Error, (Rocksky.take_string(C.rocksky_last_error()) || "connect failed") if ptr.null?

      new(ptr)
    end

    def initialize(ptr)
      @ptr = ptr
      @handlers = {}
      @thread = nil
    end

    # Register a handler for a server event type, one of +:devices+,
    # +:device_registered+, +:device_unregistered+, +:primary_changed+,
    # +:now_playing+, +:status+, +:queue+. The block receives the full
    # symbol-keyed event Hash. A +:now_playing+ event carries the advertised
    # track under +:track+, including optional +:codec+ / +:sampleRate+ audio
    # info when the player supplied it. Returns +self+ for chaining.
    def on(event_type, &block)
      @handlers[event_type.to_sym] = block
      self
    end

    # Block until the next update, returned as a symbol-keyed Hash (each carries a
    # +:type+), or +nil+ once disconnected.
    def next_event
      return nil if @ptr.null?

      ev = Rocksky.unwrap(C.rocksky_remote_controller_next_event(@ptr))
      ev && Rocksky.symbolize(ev)
    end

    # Spawn a background thread that polls #next_event and dispatches each event
    # to its registered handler until the controller disconnects. Returns the
    # Thread.
    def listen
      @thread ||= Thread.new do
        loop do
          ev = next_event
          break if ev.nil?

          dispatch(ev)
        end
      end
    end

    # Choose the primary (scrobble/profile) device.
    def set_primary(device_id)
      Rocksky.unwrap(C.rocksky_remote_controller_set_primary(@ptr, device_id.to_s))
    end

    # +play+. +target+ is a device id, or omit/nil to broadcast to all devices.
    def play(target = nil)
      command("play", target)
    end

    def pause(target = nil)
      command("pause", target)
    end

    def next(target = nil)
      command("next", target)
    end

    def previous(target = nil)
      command("previous", target)
    end

    # Send a simple command ("play" | "pause" | "next" | "previous").
    def command(action, target = nil)
      Rocksky.unwrap(C.rocksky_remote_controller_command(@ptr, action.to_s, target.to_s))
    end

    # Seek +target+ to +position_ms+ (nil target = broadcast).
    def seek(target, position_ms)
      Rocksky.unwrap(C.rocksky_remote_controller_seek(@ptr, target.to_s, position_ms))
    end

    # Jump to +index+ in +target+'s queue (nil target = broadcast).
    def queue_jump(target, index)
      Rocksky.unwrap(C.rocksky_remote_controller_queue_jump(@ptr, target.to_s, index))
    end

    # Remove queue item +index+ on +target+ (nil target = broadcast).
    def queue_remove(target, index)
      Rocksky.unwrap(C.rocksky_remote_controller_queue_remove(@ptr, target.to_s, index))
    end

    # Move queue item +from+ to position +to+ on the target device (arrayMove
    # semantics; empty target broadcasts).
    def queue_move(target, from, to)
      Rocksky.unwrap(C.rocksky_remote_controller_queue_move(@ptr, target.to_s, from, to))
    end

    # Turn queue shuffle on/off (empty target broadcasts). A player without
    # shuffle ignores it.
    def set_shuffle(target, enabled)
      Rocksky.unwrap(C.rocksky_remote_controller_set_shuffle(@ptr, target.to_s, enabled ? 1 : 0))
    end

    # Set the queue repeat mode: "off" | "all" | "one" (empty target
    # broadcasts). A player without repeat ignores it.
    def set_repeat(target, mode)
      Rocksky.unwrap(C.rocksky_remote_controller_set_repeat(@ptr, target.to_s, mode.to_s))
    end

    # Set output volume 0.0..=1.0 (empty target broadcasts). A player without
    # volume control ignores it.
    def set_volume(target, volume)
      Rocksky.unwrap(C.rocksky_remote_controller_set_volume(@ptr, target.to_s, volume.to_f.clamp(0.0, 1.0)))
    end

    # Apply a partial audio-settings document — +settings_json+ is the JSON of
    # the protocol's +audio_settings+ args (PROTOCOL.md §6.1). Raises on
    # invalid JSON rather than silently sending nothing.
    def set_audio_settings(target, settings_json)
      Rocksky.unwrap(C.rocksky_remote_controller_set_audio_settings(@ptr, target.to_s, settings_json.to_s))
    end

    # Enqueue +tracks+ (array of camelCase queue-item Hashes) on +target+. +mode+
    # is "now" | "next" | "last"; nil target = broadcast.
    def enqueue(target, tracks, mode: "now", shuffle: false, start_index: 0)
      Rocksky.unwrap(C.rocksky_remote_controller_enqueue(
                       @ptr, target.to_s, JSON.generate(tracks), mode.to_s, shuffle ? 1 : 0, start_index
                     ))
    end

    # Disconnect and stop the background task. #next_event then returns nil, which
    # ends the #listen thread. The handle stays valid until #close.
    def disconnect
      Rocksky.unwrap(C.rocksky_remote_controller_disconnect(@ptr))
    end

    # Release the native handle. The controller is unusable afterwards.
    def close
      return if @ptr.null?

      # Stop the background task so a blocked #next_event returns nil and the
      # #listen loop exits, then WAIT for that thread to leave the native call
      # before freeing — freeing the handle while a poll is in flight is a
      # use-after-free (segfault).
      C.rocksky_remote_controller_disconnect(@ptr)
      @thread&.join unless @thread == Thread.current
      @thread = nil
      C.rocksky_remote_controller_free(@ptr)
      @ptr = Fiddle::Pointer.new(0)
    end

    private

    def dispatch(event)
      handler = @handlers[event[:type]&.to_sym]
      handler&.call(event)
    end
  end
end
